drivers/ethernet/eth_xlnx_gem.c

```diff
@@ -0,0 +1,1733 @@
+ *             in the DMA memory area,
+ *         (3) the transmission completion notification timing out
+ * @retval 0 if the packet was transmitted successfully
+ */
+static int eth_xlnx_gem_send(const struct device *dev, struct net_pkt *pkt)
+{
+	const struct eth_xlnx_gem_dev_cfg *dev_conf = DEV_CFG(dev);
+	struct eth_xlnx_gem_dev_data *dev_data = DEV_DATA(dev);
+
+	uint16_t tx_data_length = 0;
+	uint16_t tx_data_remaining = 0;
+	void *tx_buffer_offs = NULL;
+
+	uint8_t bds_reqd = 0;
+	uint8_t curr_bd_idx = 0;
+	uint8_t first_bd_idx = 0;
+
+	uint32_t reg_val = 0;
+	int sem_status = 0;
+
+	if (!dev_data->started ||
+		net_if_flag_is_set(dev_data->iface, NET_IF_UP) == false ||
+		dev_data->eff_link_speed == LINK_DOWN) {
+#ifdef CONFIG_NET_STATISTICS_ETHERNET
+		dev_data->stats.tx_dropped++;
+#endif
+		return -EIO;
+	}
+
+	tx_data_length = tx_data_remaining = net_pkt_get_len(pkt);
+	if (tx_data_length == 0) {
+		LOG_ERR("%s cannot TX, zero packet length",
+			dev->name);
+#ifdef CONFIG_NET_STATISTICS_ETHERNET
+		dev_data->stats.errors.tx++;
+#endif
+		return -EINVAL;
+	}
+
+	/*
+	 * Check if enough buffer descriptors are available for the amount
+	 * of data to be transmitted, update the free BD count if this is
+	 * the case. Update the 'next to use' BD index in the TX BD ring if
+	 * sufficient space is available. If TX done handling, where the BD
+	 * ring's data is accessed as well, is performed via the system work
+	 * queue, protect against interruptions during the update of the BD
+	 * ring's data by taking the ring's semaphore. If TX done handling
+	 * is performed within the ISR, protect against interruptions by
+	 * disabling the TX done interrupt source.
+	 */
+	bds_reqd = (uint8_t)((tx_data_length + (dev_conf->tx_buffer_size - 1))
+		/ dev_conf->tx_buffer_size);
+
+	if (dev_conf->defer_txd_to_queue) {
+		k_sem_take(&(dev_data->txbd_ring.ring_sem), K_FOREVER);
+	} else {
+		sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IDR_OFFSET);
+	}
+
+	if (bds_reqd > dev_data->txbd_ring.free_bds) {
+		LOG_ERR(
+			"%s cannot TX, packet length %hu requires "
+			"%hhu BDs, current free count = %hhu",
+			dev->name, tx_data_length, bds_reqd,
+			dev_data->txbd_ring.free_bds);
+
+		if (dev_conf->defer_txd_to_queue) {
+			k_sem_give(&(dev_data->txbd_ring.ring_sem));
+		} else {
+			sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+				dev_conf->base_addr + ETH_XLNX_GEM_IER_OFFSET);
+		}
+#ifdef CONFIG_NET_STATISTICS_ETHERNET
+		dev_data->stats.tx_dropped++;
+#endif
+		return -EIO;
+	}
+
+	curr_bd_idx = first_bd_idx = dev_data->txbd_ring.next_to_use;
+	dev_data->txbd_ring.next_to_use =
+		(first_bd_idx + bds_reqd) % dev_conf->txbd_count;
+	dev_data->txbd_ring.free_bds -= bds_reqd;
+
+	if (dev_conf->defer_txd_to_queue) {
+		k_sem_give(&(dev_data->txbd_ring.ring_sem));
+	} else {
+		sys_write32(ETH_XLNX_GEM_IXR_TX_COMPLETE_BIT,
+			dev_conf->base_addr + ETH_XLNX_GEM_IER_OFFSET);
```

Comment:
and here. Actually all sys_writexx() calls seem to have these indentation issue, could you check them all in one go.